disable cross section loading on plot mode

bam241 opened the pull request:
This should disable cross section loading when on plot mode.

this also assumes that `settings::photon_transport` and `settings::run_CE` are set to `False` by default.

For a clean implementation, one could add in `Material.cpp` the `settings::run_mode != RunMode::PLOTTING` condition line 248 and line 309 to ensure the corresponding data lib are not required.

bam241 commented:
> this also assumes that settings::photon_transport and settings::run_CE are set to False by default.


I guess the CI answered my question... `settings::run_CE` can be set at True during plot mode

bam241 commented:
an other way to tackle this would be to:
we could set to False `settings::run_CE` and `settings::photon_transport` flags in PLot mode, and we could define a `settings::neutron_transport` variable that is True by default and False in plot mode

gridley commented:
I really like the `settings::neutron_transport` idea, as more people are coming to openmc that are interested only in photon problems. Why make them load neutron XS every time?

gridley commented:
Just came across another time when this is useful. If analyzing depletion results, you are pretty much stuck with calling `openmc_find_cell` from python a bunch of times to make a plot of say, nuclide densities in a slice. This is of course not somewhere where we need any XS loaded again, so it would be super nice to also have a way to turn off loading the XS when interfacing with the C API for purely geometric operations.

paulromano reviewed (commented):
I love the idea here and this is something I've personally wanted for a while. The changes here may work for the regular plotting mode, but unfortunately they would break the [openmc-plotter](https://github.com/openmc-dev/plotter/) application (which tends to be a lot more useful than generating individual plots through the API). The general problem is that we use atomic masses in order to normalize material and nuclide densities, which are displayed to the user as part of the plotter application. There are some potential ways around this, but they will require a little bit more reworking than what's present in this PR. The options that come to mind are:

- Some default atomic masses could be stored on the C++ side of the code and used when data is not loaded.
- The Python code could cache the atomic masses in some global application cache and only loaded data from the .h5 files when the cache is empty/dirty.
- The C++ code could still open the data files but _only_ load the atomic masses, which presumably would be way faster than loading all the cross section data.

Can anybody else think of other options? (ping @pshriwise)

gridley commented:
@paulromano which C API function is being called here? Is the problem that `openmc.data.atomic_mass` is not accessible in C++? Is there some way to refactor the interface calling code on the python side to use `openmc.data.atomic_mass`?

bam241 commented:
Thank you @paulromano for your feedback.

I was wondering could we use [this](https://www-nds.iaea.org/amdc/ame2020/mass_1.mas20.txt) or something equivalent if provided, and fall back on actual behavior if not provided.
This will allow a fast loading if required, but still allow folks to their usual workflow without change if their were to made aware of this capability ?

paulromano commented:
Looking through the code a bit, I realized I may have misinterpreted what the PR was meant to achieve. I had forgotten that in our plotting mode, it is already true that we don't load cross sections from .h5 files. It looks like this PR is meant to avoid even touching cross_sections.xml. Is that correct @bam241?

Another point of clarification -- with respect to the plotter, we actually don't load OpenMC in plotting mode but rather in volume calculation mode (which does require loading atomic masses). If we want to speed up data loading for the plotter (and volume calculations), I think the easiest path is my last suggestion from above, load only the masses from the .h5 files, which should be very quick. That change could be made separate from this PR.

shimwell commented:
I think it would be nice is users don't need to have a cross_sections.xml when plotting.

Also I think it would be nice if we can avoid getting the atomic data from the cross section h5 files when plotting

Perhaps we can get to this point by building on this PR is the manner @gridley suggested.

What do people think is this something?

pshriwise commented:
Just had a chance to read through the discussion above.

As @paulromano said, we've been wanting to remove the cross section data requirement when using OpenMC for visualization purposes. My understanding is that by skipping loading all data other than the atomic weight ratios from the cross section files we could do this.

As far as removing the `cross_sections.xml` requirement goes, I'm wondering if this would be a path forward:

  - convert the `mass_1.mass20.txt` file to another format readable from Python or C++ with our current dependencies. Ideally, I'd say that this should be HDF5, but since it's a file we keep in our repo maybe XML would be a better choice. Let's pretend we go with XML from here.
  - Transition `openmc.data.atomic_mass` to read from the XML file.
  - When initializing OpenMC in plotting mode:
    - if a `cross_section.xml` data is available, load information from the HDF5 data (AWR only)
    - if no `cross_section.xml` data is available, load information from the `atomic_mass.xml` file. (We'll potentially want to print a warning here indicating that the masses aren't loaded from the cross section data.)

A couple of additional thoughts/open questions:
  - The script used to convert the `.txt` file to another format should be added

While I think this is pretty clean in terms of the data management, one nontrivial roadblock to something like this is reliably indicating to the `openmc` executable where to find the atomic mass data... I'll think on that a bit.

bam241 commented:
Just wondering, why not hard code it into the cpp, and using the h5 atomic mass by default, and use the hardcoded one by default on plot mode ?

pshriwise commented:
> Just wondering, why not hard code it into the cpp, and using the h5 atomic mass by default, and use the hardcoded one by default on plot mode ?

Yeah I thought about that too. Put it in a header file or something. Just would be nice to have one source for both Python and C++. Definitely open to other thoughts.

bam241 commented:
or openMC could ship a data file with those ?

pshriwise commented:
> or openMC could ship a data file with those ?

The resulting XML file I referenced above would be shipped with OpenMC just like the text file we use now on the Python side 👍🏻 

> but there is something I am missing, the openMC python is tied with the openMC C++, so if we add it into the c++ should it not be accessible through python ?

The C++ library is accessible (to some degree) using the `openmc.lib` module, but what you get with the primary Python API (`import openmc`) is a pure Python interface that doesn't interact with the compiled code.

paulromano commented:
Another option came to mind -- in the case that the user wants to do plotting through the plotter GUI and they don't have cross section data set up, we could just _not_ provide any density information. I think everything else would still work.

pshriwise commented:
> Another option came to mind -- in the case that the user wants to do plotting through the plotter GUI and they don't have cross section data set up, we could just _not_ provide any density information. I think everything else would still work.

True. Maybe load from HDF5 if a `cross_sections.xml` is provided and if not, no densities in the plotter. Or maybe a setting?

bam241 commented:
Thank you all for the interest, 
@paulromano @pshriwise to simplify this, should we add a flag to allow to avoid reading `cross_section.xml` ? this would requires enabling an extra flag, we won't break compatibility, 
further change could allow to have this flag also enable when no cross_section.xml is found (in plot mode) and have some repercussion of the `openmc-plotter` capabilities ?

bam241 commented:
@paulromano @pshriwise I added guards to allow volume calculations without cross_section.xml file, if not present this should skip atom density calculation.
I implemented the guard so it check if `settings::path_cross_sections` is set, that might not be the best way, but I didn't want to add an additional boolean.

I didn't add tests either. (that might be needed to ensure this works properly...)

If you add a better implementation in mind, or a better guards system, let me know.

I can also remove this change from the PR and have it implemented an other time.

paulromano commented:
@bam241 I made a few changes here on your branch and just want to make sure you agree with them:
- I've moved the conditional check on plotting mode from within `read_cross_sections_xml` to outside it. The way you had it before, if someone had a model.xml file with plots specified inside, it would still end up loading cross_sections.xml.
- In the `Material` constructor, I've limited the scope of the conditional to just the part where a `LibraryKey` is constructed.
- The check you had at the beginning of `openmc_load_nuclide` didn't seem necessary because it shouldn't be called in plotting model, so I removed that.

bam241 commented:
> * I've moved the conditional check on plotting mode from within `read_cross_sections_xml` to outside it. The way you had it before, if someone had a model.xml file with plots specified inside, it would still end up loading cross_sections.xml.
> * In the `Material` constructor, I've limited the scope of the conditional to just the part where a `LibraryKey` is constructed.

This sounds sounds good, I was unsure of the best approach to do this. I am happy to shimmed in!


> * The check you had at the beginning of `openmc_load_nuclide` didn't seem necessary because it shouldn't be called in plotting model, so I removed that.

A left over debugging when I was trying to have the Volume calculation works with and without the `cross_section.xml`. I started to loose myself in the conditional, so I decided to do separated PR for this.

paulromano reviewed (approved):
All good to go now. Thanks @bam241!
